library/src/blas1/scal_host.hpp

```diff
@@ -0,0 +1,82 @@
+                                             const U*       alpha,
+                                             T*             x[],
+                                             rocblas_int    offsetx,
+                                             rocblas_int    incx,
+                                             rocblas_int    batch_count)
+{
+    static constexpr int NB = 256;
+    // Quick return if possible. Not Argument error
+    if(n <= 0 || incx <= 0 || batch_count <= 0)
+        return rocblas_status_success;
+
+    dim3        blocks((n - 1) / NB + 1, batch_count);
+    dim3        threads(NB);
+    hipStream_t rocblas_stream = handle->rocblas_stream;
+
+    if(rocblas_pointer_mode_device == handle->pointer_mode)
+        hipLaunchKernelGGL(
+            scal_kernel_batched, blocks, threads, 0, rocblas_stream, n, alpha, x, offsetx, incx);
+    else // alpha is on host
+        hipLaunchKernelGGL(
+            scal_kernel_batched, blocks, threads, 0, rocblas_stream, n, *alpha, x, offsetx, incx);
+
+    return rocblas_status_success;
+}
+
+template <typename T, typename U>
+rocblas_status rocblas_scal_strided_batched_template(rocblas_handle handle,
+                                                     rocblas_int    n,
+                                                     const U*       alpha,
+                                                     T*             x,
+                                                     rocblas_int    offsetx,
+                                                     rocblas_int    incx,
+                                                     rocblas_int    stridex,
+                                                     rocblas_int    batch_count)
+{
+    static constexpr int NB = 256;
+    // Quick return if possible. Not Argument error
+    if(n <= 0 || incx <= 0 || batch_count == 0)
+        return rocblas_status_success;
+
+    dim3        blocks((n - 1) / NB + 1, batch_count);
+    dim3        threads(NB);
+    hipStream_t rocblas_stream = handle->rocblas_stream;
+
+    if(rocblas_pointer_mode_device == handle->pointer_mode)
+        hipLaunchKernelGGL(scal_kernel_strided_batched,
+                           blocks,
+                           threads,
+                           0,
+                           rocblas_stream,
+                           n,
+                           alpha,
+                           x,
+                           offsetx,
+                           incx,
+                           stridex);
+    else // alpha is on host
+        hipLaunchKernelGGL(scal_kernel_strided_batched,
+                           blocks,
+                           threads,
+                           0,
+                           rocblas_stream,
+                           n,
+                           *alpha,
+                           x,
+                           offsetx,
+                           incx,
+                           stridex);
+
+    return rocblas_status_success;
```

Comment:
Why is `scal` implementation duplicated between batched and non-batched? Couldn't non-batched just be expressed with a `bool` template parameter (before any type parameters which can be deduced automatically), so that code is not duplicated and 90% of the batched and non-batched code is shared? With C++17 you can use `constexpr if`, but we're still using C++14, but as long as both halves of an `if` with a constant template parameter are compileable, the compiler optimizes away the false one.

Reply:
The non-batched version currently uses the strided_batched call at the template layer. I have different templates for the batched version and strided_batched version. After talking with Juan, I can move these to a single template file using the preprocessor.

Reply:
Created one shared rocblas_scal_template function using the preprocessor to differentiate batched vs strided_batched/non-batched calls. (see caacbf0)